File: c++/Solution40.cpp
//Backtracking Knapsack O(n * 2^n) O(target), n = candidates.size()
class Solution40 {
public:
    vector<vector<int>> res;

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end()); // 要避免重複, 所以排序小的先入列篩選

        vector<int> selected;
        backtracking(0, selected, candidates, target);
        return res;
    }

    void backtracking(int start, vector<int>& selected, vector<int>& candidates, int target) {
        if (target == 0) {
            res.push_back(selected);
            return;
        }
        if (target < 0) return;

        for (int i = start; i < candidates.size(); ++i) {
            if (i > start && candidates[i] == candidates[i - 1]) continue; // 跳過重複的, 因為該項已經在前一個情況出現過, 即只挑選candidates[i]的那個情況

            selected.push_back(candidates[i]);
            backtracking(i + 1, selected, candidates, target - candidates[i]);
            selected.pop_back();
        }
    }
};
/* sort candidates = [1,2,2,2,5], target = 5, sort
_ -> 1 -> 12 -> 122(o)
                122(skip)
                125(x)
          12(skip, 如果繼續做, 相當於上上面的那個122(skip))
          12(skip)
          15(x)
     2 -> 22 -> 222(x)
                225(x)
          22(skip)
          25(x)
     2(skip)
     2(skip)
     5(0)

res=[[1,2,2],[5]]
*/

File: java/Solution40.java
package leetCode.java;

import java.util.*;

//Backtracking Knapsack O(n * 2^n) O(target), n = candidates.size(), DFS每次new ArrayList需要O(n), 共O(n * 2^n), 此時間複雜度大於排序所需
class Solution40 {
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates); // 排序後相同元素會相鄰，方便跳過重複組合
        backtracking(0, new ArrayList<>(), candidates, target);
        return res;
    }

    public void backtracking(int start, List<Integer> selected, int[] nums, int target) {
        if (target == 0) {
            res.add(new ArrayList<>(selected));
            return;
        }
        if (target < 0) {
            return;
        }

        for (int i = start; i < nums.length; ++i) {
            if (i > start && nums[i] == nums[i - 1]) {
                continue; // 一樣的數字的話要剪掉, 否則backtracking會再處理一次
            }
            selected.add(nums[i]);
            backtracking(i + 1, selected, nums, target - nums[i]);
            selected.remove(selected.size() - 1);
        }
    }
}